### PR TITLE
Improve markdown editor: width, height, preferred

### PR DESCRIPTION
**Pull request by wxiaoguang**
Follow #23876

1. Fine tune the heights of the editors (like before)
    * Auto expand the editor (increase/decrease the height) when editing
2. Remember user's last used editor (textarea/easymde) in LocalStorage, then next time the editor will be switched automatically
    * No need to introduce extra config option, it satisfies all users, including who prefer EasyMDE
3. Also fix the width problem of Review Panel

Screenshot:

<details>

image

image

image

image

image

image

</details>

### CODE REVIEW

File: web_src/js/features/repo-wiki.js
Comment (silverwind) [inline web_src/js/features/repo-wiki.js:51]: Maybe `height: min(calc(100vh - 600px), 300px)`? I recently came to like `min` and `max` functions because they work on all properties. Also, I think in many cases, they can replace media queries.

Ideally I think this should be set from CSS, not JS.

Comment (wxiaoguang) [reply web_src/js/features/repo-wiki.js:51]: EasyMDE has a bug, it has inline `style`, so we have to overwrite it by JS.

I will add a comment.

ps: I think this `editorHeights` is clear enough, it is a clear definition and won't cause conflicts.

Comment (wxiaoguang) [reply web_src/js/features/repo-wiki.js:51]: Updated in ab1b645a62a0036533bb1b6f483c21a648d4938d

More details:

https://github.com/Ionaru/easy-markdown-editor/issues/413

https://github.com/Ionaru/easy-markdown-editor/blob/d26b4e399b42c8b209531018393454cc1bdd5f6c/src/js/easymde.js#L1857

Comment (silverwind) [reply web_src/js/features/repo-wiki.js:51]: Ok, we should move this to CSS once we remove EasyMDE.

File: web_src/js/features/repo-wiki.js
Comment (silverwind) [inline web_src/js/features/repo-wiki.js:49]:
```suggestion
    // And another benefit is that we only need to write the style once for both editors.
    // TODO: Move height style to CSS after EasyMDE removal.
```

File: web_src/js/features/comp/ComboMarkdownEditor.js
Comment (silverwind) [inline web_src/js/features/comp/ComboMarkdownEditor.js]: Maybe outsource this to https://github.com/jackmoore/autosize? This stuff is tricky to get right.

Comment (wxiaoguang) [reply web_src/js/features/comp/ComboMarkdownEditor.js]: It uses `getComputedStyle`, adds many event listeners, while I don't see new benefit it brings.

If in some cases our code has bugs or it works better than our code, we can introduce it. Otherwise, I think the current code is simple enough and just works well.

Comment (silverwind) [reply web_src/js/features/comp/ComboMarkdownEditor.js]: Yeah I guess if we can do without, it'll be better. `getComputedStyle` is probably unnecessary if we know the padding/line-height/font-size beforehand.

I recall I've implemented something similar on another project (it uses React but the code is same for native). On `input` event, trigger below code. The values depend on styling of the element so will be different for us.

```js
el.style.height = "auto";
const maxHeight = 17.6/* padding */ + 16 * 1.4 * maxRows/* row with line-height */;
const height = el.scrollHeight;
el.style.height = `${Math.min(height, maxHeight)}px`;
el.style.maxHeight = `${maxHeight}px`;
if (height <= maxHeight) {
  el.style.overflowY = "hidden";
} else {
  el.style.removeProperty("overflow-y");
}
```

Comment (wxiaoguang) [reply web_src/js/features/comp/ComboMarkdownEditor.js]: Yup, the ideas are all the same: set `height=auto`, get new `scrollHeight`, decide how to set new height or scroll.

File: web_src/css/editor-markdown.css
Comment (silverwind) [inline web_src/css/editor-markdown.css:27]: We should allow manual resize. If manual resize was done, don't auto resize any more.

Comment (silverwind) [inline web_src/css/editor-markdown.css]:
```suggestion
  resize: vertical;
```


We should allow manual resize. If manual resize was done, don't auto resize any more.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: Done in 9af6a4489bb68c5b220282f9af44cf7621209c2a

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Wouldn't allow horizontal resize.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: Sorry I didn't get the point.

You can take a try, the current behavior is good enough. If you think anything is wrong, please help to show an example.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Just set `resize: vertical` which will ensure it will only resize vertically. I can test later. I'm not sure if `ResizeObserver` is really necessary. Just set a boolean when the user ever touched the resize, and don't do any automatic resizing after that.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: I think we do not need to set `resize`.

`ResizeObserver` is necessary, otherwise, if user resize the textarea to a larger size, then they change the content value, then the textarea will shrink to the initial height. By `ResizeObserver`, we can remember that the new height as initial height, then the textarea won't shrink smaller that that.

"Just set a boolean when the user ever touched the resize": it still needs `ResizeObserver` IMO.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Pretty certain boolean will be enough.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: > Pretty certain boolean will be enough.

At the moment, I do not know how to distinguish "user resizing" and "browser resizing" behaviors accurately (even if it could be done, it needs more code).

If you use a boolean, it might be set to true by browser resizing, then the user won't have the "auto resizing"  anymore.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: I guess there must be some property on the resize event to distinguish.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: We could also just use https://github.com/github/textarea-autosize which github uses. Better than to re-invent the wheel.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: AFAIK, I haven't found one. 

https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Above repo uses `mousemove` to detect user resize. Let's just integrate that module please.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: That's much more complicated than it should be. I really do not like it.

And it uses the `getComputedStyle` which you have expressed strong objection.

----

Update: I can't even see why the current "resize" behavior could be a problem. Feel free to try.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Then at least implement the mousemove detection. Also check if pasting correctly resizes.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: Actually I like current behavior. I am not sure I can get  100% your point. If you want to change the behavior, feel free to edit this PR directly (I can accept either resize behavior, they don't bother me)

Comment (silverwind) [reply web_src/css/editor-markdown.css]: I find `ResizeObserver` needlessly complicated and potentially error-prone. I would pretty much implement it like that module, minus the `getComputedStyle`, just a simple boolean to track user interaction.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: Well, I can not fully understand your point. The `autosize` says: `future todo: consider replacing with ResizeObserver`

https://github.com/jackmoore/autosize/blob/1e0229b9bb9f51c2b6a2756d901a023b475fa950/src/autosize.js#L130


`ResizeObserver` is correct and it is the future.

So, feel free to edit this PR for your idea.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: If you use it, you should at minimum debounce the callback as this will trigger hundrets of times per second and consume unnecessary CPU.

Still I see no need, boolean with mousemove is simple enough. Should test if it triggers for touch resize too, but I assume it would.

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: > If you use it, you should at minimum debounce the callback as this will trigger hundrets of times per second and consume unnecessary CPU.

I added `console.log`, I didn't see so. We only obverse on the textarea.

`mousemove` could also be triggered plenty of times when moving mouse.

----

To resolve the disagreement, I added `auotsize` package. And leave a `TODO` in comment: 

```
  resize: vertical; /* TODO: even if we set this, the `autosize` package doesn't support manually resize vertically */
```

----



What's your opinion? Continue using the autosize package? Or I use my ResizeObserver (which has no performance problem at all) to improve?

Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: What's your opinion? Continue using the `autosize` package? Or I use my ResizeObserver (which has no performance problem at all) to improve?

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Will test. Surprised about the TODO because GitHub UI certainly allows manual resize and uses that package.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Ah, no, only `@github/textarea-autosize` supports the manual resize.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: I tried `@github/textarea-autosize` package but it shows a different bug where the cursor leaves no padding at the bottom. It also reproduces on GitHub.

Screenshot 2023-04-05 at 22 05 41

One nice feature of the package is their `viewportMarginBottom` option, which when set to a value like 130 will ensure the comment button never goes offscreen during resize:

Screenshot 2023-04-05 at 22 09 44

So I'd suggest either switching to to `@github/textarea-autosize` and live with the cursor padding bug, or go back to the `ResizeObserver` and try to re-implement the `viewportMarginBottom` behaviour while ensuring the size is calculated correctly so the bottom-padding is visible.

**Comment by silverwind** [comment]
I'll test the ResizeObserver method as well by reverting the last commit locally, hold on...

File: web_src/css/editor-markdown.css
Comment (silverwind) [reply web_src/css/editor-markdown.css]: Actually I think I will just incorporate that function from the module into our code. It does a few nice things over the `ResizeObserver` method, like downsizing as well when content is deleted. I think I should be able to work out that padding bug, than it'll be perfect.

Comment (silverwind) [reply web_src/css/editor-markdown.css]: Pushed up my version of the `autosize` function. Changes include:

- Removal of `getComputedStyle` of parent element, no observable change in functionality
- Add scrolling to bottom when cursor is at the end after an input event. This prevents the padding issue above which is caused by the textarea not completely scrolled.
- Changed style unsets to use `removeProperty` instead of setting to empty string.

If we could also get rid of the `getComputedStyle` on the textarea, it would be perfect, but it might not be easy so this is LGTM with it too.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Regarding licensing of the modified module, I've put the license inline to satisfy MIT license. We could also opt to include it in `licenses.txt`, but for that I'd need to devise add a mechanism to add licenses for things that we do not depend upon.

File: web_src/css/editor-markdown.css
Comment (wxiaoguang) [reply web_src/css/editor-markdown.css]: > It does a few nice things over the ResizeObserver method, like downsizing as well when content is deleted

ResizeObserver  could also do so without bug, I didn't implement it correctly before, and it doesn't have performance problem.

Since you have implemented the "mouse*" event listener for resize, then there is no need to change, I am fine with it.

File: web_src/js/utils/dom.js
Comment (wxiaoguang) [inline web_src/js/utils/dom.js]: I guess these two events are redundant. Only `input` is enough

File: web_src/js/utils/dom.js
Comment (wxiaoguang) [inline web_src/js/utils/dom.js]: I guess we couldn't remove these styles, instead, as a framework library, it should backup the initial values and reset to initial values.

File: web_src/js/utils/dom.js
Comment (wxiaoguang) [inline web_src/js/utils/dom.js]: Doesn't pair with above (no `paste` event)? While I guess we do  not need to handle the keyup/paste events, only `input` is enough

File: web_src/js/utils/dom.js
Comment (silverwind) [reply web_src/js/utils/dom.js]: Might be, need to test more browsers.

File: web_src/js/utils/dom.js
Comment (silverwind) [reply web_src/js/utils/dom.js]: Possible. Original code did set to empty string which I think is equal to current.

**Comment by wxiaoguang** [comment]
Made some changes:

1. Set the default value of `viewportMarginBottom` to 0, instead of the magic number 100. The `100` doesn't look like a general number. The callers should fine tune the margin by themselves, instead of assuming a magic number.
2. Introduce `backupStyle` and `restoreStyle` to help to reset to the original styles.
3. Rename `height` to `styleHeight`, it is not a number, but something like `100px`.
4. Remove `textarea.removeEventListener('mousemove', onUserResize);`, because you want to do `reset`, if you do reset after the listener removal, then the `onUserResize` logic doesn't work anymore.
5. Use `scrollY` instead of `pageYOffset`, because `pageYOffset` is deprecated.
6. Rename `sizeToFit` to `resizeToFit` , and expose it, because it needs to be called explicitly when the `value` changes.
7. Remove `keyup` and `paste` event listeners. Modern browsers always trigger the `input` event if the content is changed by users.

There are still 2 things which are not L-G-T-M:

1. The usage of `max-height`, I do not think we should touch it anymore. We can calculate the height accurately by our code, so we do not need to use `max-height` to limit anything. Instead, we should leave the `max-height` chance to developers, they could define different `max-height` by CSS styles.
2. A bug, if the textarea is not in viewport, then the `resizeToFit` doesn't work. The textarea never shrinks even if the lines are deleted. See the screenshot. Steps:
    1. Input many new lines, to expand the textarea.
    2. Scroll part of the textarea out of the viewport.
    3. Delete lines, the textarea never shrinks.

<details>

image

</details>

**Comment by wxiaoguang** [comment]
I pushed more changes 5bddb63b7b1d641a4e9be4b89cb2351cd66d0f2b , I think these 2 problems get resolved.

**Comment by silverwind** [comment]
Seems some good changes. I will give this another test soon.

> Remove textarea.removeEventListener('mousemove', onUserResize);, because you want to do reset, if you do reset after the listener removal, then the onUserResize logic doesn't work anymore.

I see you added it back. I think the `destroy` function should cleanup all event subscriptions, not leave some subscribed.

**Comment by wxiaoguang** [comment]
> Seems some good changes. I will give this another test soon.
> 
> > Remove textarea.removeEventListener('mousemove', onUserResize);, because you want to do reset, if you do reset after the listener removal, then the onUserResize logic doesn't work anymore.
> 
> I see you added it back. I think the `destroy` function should cleanup all event subscriptions, not leave some subscribed.

I mean this one, I removed this `removeEventListener()` line

image

**Comment by silverwind** [comment]
Not sure about that change. I think it was so that after the first user-initiated event, `isUserResized` becomes true so all automatic resizing is stopped, and in which case it does not make sense to listen for any more events.

I think it made sense, just need to check that the double removal during `destroy` does not error when the listener was already removed.

**Review by silverwind — APPROVED** [review]
Works well.

**Comment by wxiaoguang** [comment]
Do we still need to keep the LICENSE comment? I think the code is totally  different ......

**Review by lunny — APPROVED** [review]

**Comment by silverwind** [comment]
Depens on [interpretation](https://opensource.stackexchange.com/a/2188) of the license, I would keep it.

**Comment by wxiaoguang** [comment]
So this can be `wait-merge` now?

**Comment by silverwind** [comment]
I don't know if I'm allowed to set this tag 😆

**Comment by delvh** [comment]
I think everyone/ every maintainer is allowed to set it given the following conditions are met:
- The PR has at least two approvals
- There are no open discussions
- There are no merge conflicts that need to be resolved

**Comment by jolheiser** [comment]
🎺 🤖

**Review by jolheiser — APPROVED** [review]

**Review by jolheiser — APPROVED** [review]

**Pull request merged**
